hyperspy/io_plugins/nexus.py

```diff
@@ -0,0 +1,1251 @@
+                    'offset': 0.0,
+                    'units': '',
+                    'navigate': True,
+                   })
+        if signal_type == "Signal2D":
+            nav_list[-1]["navigate"]=False
+            nav_list[-2]["navigate"]=False
+        elif signal_type == "Signal1D":
+            nav_list[-1]["navigate"]=False
+            
+    title = _text_split(nexus_dataset_path, '/')[-1]
+    metadata = {'General': {'title': title}} 
+    dictionary = {'data': data_lazy,
+                  'axes': nav_list,
+                  'metadata':metadata}
+    return dictionary
+
+
+def file_reader(filename,lazy=True, dataset_keys="hardlinks",
+                metadata_keys="all",
+                nxdata_only=True,
+                small_metadata_only=False,
+                **kwds):   
+    """ Reads NXdata class or hdf datasets from a file and returns signal
+    
+    Note
+    ----
+    Loading all datasets can result in a large number of signals
+    Please review your datasets and use the dataset_keys to target
+    the datasets of interest. 
+    value and keys are special keywords and prepended with fix_ in the metadata
+    structure to avoid any issues.
+    
+    Datasets are all arrays with size>2 (arrays, lists)
+    
+    Parameters
+    ----------
+    myDict : dict or h5py.File object
+    dataset_keys  : {"all","hardlinks",str,list of string} 
+        if "all" - all datasets are returned
+        if "hardlinks" - softlinks are ignored
+        Only return items whose path contain the strings
+        .e.g search_list = ["instrument","Fe"] will return
+        data entries with instrument or Fe in their hdf path.
+    metadata_keys: : str list or None
+        Only return items from the original metadata whose path contain the strings
+        .e.g search_list = ["instrument","Fe"] will return
+        all metadata entries with instrument or Fe in their hdf path.
+    nxdata_only : bool  
+        If true only NXdata will be converted into a signal
+        if false NXdata and any hdf datasets not bound within the
+        NXdata sets will be loaded as signals
+    small_metadata_only : bool, default:True  
+        If true arrays of size<2
+        will be loaded into the metadata structure. If false larger 
+        arrays will be also loaded into the metadata
+    
+
+    Returns
+    -------
+    dict
+        signal dictionary or list of signal dictionaries of the form
+        {'data': data,
+         'axes' :  axes description if available
+         'metadata': metadata,
+         'original_metadata':original_metadata,
+         'mapping':mapping}
+    
+
+    See Also 
+    --------
+    * :py:meth:`~.io_plugins.nexus.read_datasets_from_file`
+    * :py:meth:`~.io_plugins.nexus.read_metadata_from_file`
+                 
+   
+    """
+    # search for NXdata sets...
+
+    mapping  = kwds.get('mapping',{})   
+    original_metadata={}
+    learning = {}
+    fin = h5py.File(filename,"r+")
```

Comment:
The readers works on any hdf or nexus file and I tend to have a mixture. Raw h5 which are then linked to within a nexus file so I omitted a strict check. As it's filtered by the nxs extension I didn't think it was needed.  From #2377 there is a need for more straightforward way to lazy load from a file.  If you can specify the reader in the load function you could generically load from hdf.

Reply:
Agree, we need to be able to specify a reader but also have a mechanism to pick a suitable reader for a given hdf5 file, in order to avoid relying on the extension. If you can add this check now, it would avoid us coming back to this later.
This will come at some point soon - at the latest when we split hyperspy.

> From #2377 there is a need for more straightforward way to lazy load from a file. If you can specify the reader in the load function you could generically load from hdf.

Not sure if I understand correctly: you are suggesting that specifying the nexus reader in `hs.load` will allow reading any hdf5 file (even if this is not a nexus file)?